src/app/public/src/modules/clipboard/clipboard.service.ts

```diff
@@ -0,0 +1,75 @@
+// adapted from ngx-clipboard
+
+import { Injectable, Renderer2, RendererFactory2 } from '@angular/core';
+import { StacheWindowRef } from '../shared';
+
+@Injectable()
+export class StacheCopyToClipboardService {
+  private copyTextArea: HTMLTextAreaElement;
+  private document: any;
+  private renderer: Renderer2;
+
+  constructor(
+    private windowRef: StacheWindowRef,
+    private rendererFactory: RendererFactory2
+  ) {
+    this.renderer = this.rendererFactory.createRenderer(undefined, undefined);
+    this.document = this.windowRef.nativeWindow.document;
+  }
+
+  public copyContent(element: HTMLElement) {
+    let copyTarget = this.getValidInputElement(element);
+    this.copyFromInputElement(copyTarget);
+  }
+
+  private copyFromInputElement(copyTarget: HTMLInputElement | HTMLTextAreaElement) {
+    this.selectTarget(copyTarget);
+    this.document.execCommand('copy');
+    this.clearSelection(copyTarget);
+  }
+
+  private getValidInputElement(element: HTMLElement): HTMLInputElement | HTMLTextAreaElement {
+
+    if (this.isValidInputElement(element)) {
+      return element as HTMLInputElement | HTMLTextAreaElement;
+    }
+
+    return this.createTextAreaFromElement(element);
+  }
+
+  private isValidInputElement(element: HTMLElement): boolean {
+    return (element instanceof HTMLTextAreaElement || element instanceof HTMLInputElement);
+  }
+
+  private selectTarget(inputElement: HTMLInputElement | HTMLTextAreaElement) {
+    this.renderer.selectRootElement(inputElement).select();
+    this.renderer.selectRootElement(inputElement).setSelectionRange(0, inputElement.value.length);
+  }
+
+  private clearSelection(inputElement: HTMLInputElement | HTMLTextAreaElement) {
+    inputElement.blur();
+    this.windowRef.nativeWindow.getSelection().removeAllRanges();
+  }
+
+  private createTextAreaFromElement(element: HTMLElement): HTMLTextAreaElement {
```

Comment:
@Blackbaud-SteveBrush left comments on the other PR:
In src/app/public/src/modules/clipboard/clipboard.service.ts:

> @@ -43,18 +51,25 @@ export class StacheClipboardService {
     this.windowRef.nativeWindow.getSelection().removeAllRanges();
   }
 
-  private createCopyTextArea() {
-    const copyTextAreaStyles = `
-      font-size: 12px;
-      border: 0;
-      padding: 0;
-      margin: 0;
-      position: absolute;
-      left: -9999px;
-      top: ${this.windowRef.nativeWindow.pageYOffset || this.document.documentElement.scrollTop}px;`;
+  private createTextAreaFromElement(element: HTMLElement): HTMLTextAreaElement {
+    if (!this.copyTextArea) {
Part of me is wondering if this HTML/CSS "stuff" could be added to the clipboard template instead. It just feels wrong to be writing HTML in the component class.

clipboard.component.html

(The variable names are inconsequential, but something more manageable like the following...)

<ng-template [ngIf]="copiedContents">
  <textarea
    class="stache-clipboard-hidden-contents"
    readonly
  >{{ copiedContents }}</textarea>
</ng-template>

Reply:
This text area is being generated by the Service to create something copy-able out of html content that isn't a text input/area. I think removing it from the Service code itself might cripple the the ability for the service to be reused later, outside of our implementation. If we are only ever going to use the service with the component itself, then I see no issue with moving it in to the component code and grabbing it programmatically as needed

Reply:
This also prevents a new `textarea` from spawning per component.

Reply:
Good points.